### PR TITLE
Remove dependence of step-39 on LocalIntegrators.

### PR DESCRIPTION
**Pull request by bangerth**
Related to #15604, #16690. The first commit, in essence, inlines the contents of the `LocalIntegrators` functions into step-39 plus some minor clean-ups. The second commit is just a C++ issue. The third commit fixes the issue in the computation of the IP penalty factor mentioned in #15604.

### CODE REVIEW

**Comment by bangerth** [comment]
Yes, totally! If you already have such a patch, feel free to post. Otherwise, how about we merge the current one which removes one kind of dependence (on the pre-canned integrators) and you build your patch on that removing the other kind of dependence (on the various other data structures)?

I would like to get this one merged before the release because that allows us to move forward with #15604. Using a better approach to integration can work on a separate time scale if necessary.

**Comment by tjhei** [comment]
Let's merge this first then.

**Review by tjhei — APPROVED** [review]

File: examples/step-39/step-39.cc [hunk 1]
```diff
@@ -94,67 +94,191 @@ namespace Step39
...
+      GeometryInfo<dim>::unit_normal_direction[dinfo1.face_number];
+    const unsigned int normal2 =
+      GeometryInfo<dim>::unit_normal_direction[dinfo2.face_number];
+    const unsigned int deg1sq = (deg1 == 0) ? 1 : deg1 * (deg1 + 1);
+    const unsigned int deg2sq = (deg2 == 0) ? 1 : deg2 * (deg2 + 1);
+
+    double penalty1 = deg1sq / dinfo1.cell->extent_in_direction(normal1);
+    double penalty2 = deg2sq / dinfo2.cell->extent_in_direction(normal2);
+    if (dinfo1.cell->has_children() && !dinfo2.cell->has_children())
+      penalty1 *= 2;
```
Comment (tjhei) [inline examples/step-39/step-39.cc:140]: I don't understand why this is different from the old code that does an XOR between the two bools. Your version is a easier to understand, though.

Comment (bangerth) [reply examples/step-39/step-39.cc:140]: Yes, I find XORs to be exceptionally difficult to read.

The difference is that in the original code, we only always updated `penalty1`, even if the second cell was refined. In my code here, I update `penalty1` above and `penalty2` below. It doesn't matter if you're on a uniform mesh, but it matters in general.

**Review by masterleinad — APPROVED** [review]

File: examples/step-39/step-39.cc (hunk 1 above)
Comment (tjhei) [reply examples/step-39/step-39.cc:140]: Oh, gotcha!

**Pull request merged**